Propagate parent policy container to local iframes

TimvdLippe opened the pull request:
This follows the rules as defined in
https://w3c.github.io/webappsec-csp/#security-inherit-csp
where local iframes (about:blank and about:srcdoc) should
initially start with the CSP rules of the parent. After
that, all new CSP headers should only be set on the
policy container of the iframe.

Part of #36437

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

servo-wpt-sync commented:
🤖 Opened new upstream WPT pull request (web-platform-tests/wpt#52194) with upstreamable changes.

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52194).

jdm commented:
```diff
diff --git a/components/script/dom/servoparser/mod.rs b/components/script/dom/servoparser/mod.rs
index 0650fde676..5e9b9d11c5 100644
--- a/components/script/dom/servoparser/mod.rs
+++ b/components/script/dom/servoparser/mod.rs
@@ -21,6 +21,7 @@ use html5ever::{Attribute, ExpandedName, LocalName, QualName, local_name, ns};
 use hyper_serde::Serde;
 use markup5ever::TokenizerResult;
 use mime::{self, Mime};
+use net_traits::policy_container::PolicyContainer;
 use net_traits::request::RequestId;
 use net_traits::{
     FetchMetadata, FetchResponseListener, Metadata, NetworkError, ResourceFetchTiming,
@@ -813,6 +814,18 @@ impl ParserContext {
             pushed_entry_index: None,
         }
     }
+
+    pub(crate) fn override_csp_list(&self, policy_container: Option<&PolicyContainer>) {
+        let Some(policy_container) = policy_container else {
+            return;
+        };
+        let Some(parser) = self.parser.as_ref().map(|p| p.root()) else {
+            return;
+        };
+        parser
+            .document
+            .set_csp_list(policy_container.csp_list.clone());
+    }
 }

 impl FetchResponseListener for ParserContext {
diff --git a/components/script/script_thread.rs b/components/script/script_thread.rs
index bb51680260..2d0d93d6c6 100644
--- a/components/script/script_thread.rs
+++ b/components/script/script_thread.rs
@@ -3609,15 +3609,18 @@ impl ScriptThread {
             None => vec![],
         };

+        let policy_container = incomplete.load_data.policy_container.clone();
         self.incomplete_loads.borrow_mut().push(incomplete);

         let dummy_request_id = RequestId::default();
         context.process_response(dummy_request_id, Ok(FetchMetadata::Unfiltered(meta)));
+
         context.process_response_chunk(dummy_request_id, chunk);
         context.process_response_eof(
             dummy_request_id,
             Ok(ResourceFetchTiming::new(ResourceTimingType::None)),
         );
+        context.override_csp_list(policy_container.as_ref());
     }

     /// Synchronously parse a srcdoc document from a giving HTML string.
@@ -3632,6 +3635,7 @@ impl ScriptThread {
         let srcdoc = std::mem::take(&mut incomplete.load_data.srcdoc);
         let chunk = srcdoc.into_bytes();

+        let policy_container = incomplete.load_data.policy_container.clone();
         self.incomplete_loads.borrow_mut().push(incomplete);

         let mut context = ParserContext::new(id, url);
@@ -3643,6 +3647,7 @@ impl ScriptThread {
             dummy_request_id,
             Ok(ResourceFetchTiming::new(ResourceTimingType::None)),
         );
+        context.override_csp_list(policy_container.as_ref());
     }

     fn handle_css_error_reporting(
```

This diff makes `tests/wpt/tests/content-security-policy/unsafe-eval/eval-blocked-in-about-blank-iframe.html` pass for me. It makes sense, because about:blank and about:srcdoc don't go through the network, so the RequestBuilder doesn't get used there. There's a similar setup for propagating the referrer policy from the parent document as well.

TimvdLippe committed:
Propagate parent policy container to local iframes

This follows the rules as defined in
https://w3c.github.io/webappsec-csp/#security-inherit-csp
where local iframes (about:blank and about:srcdoc) should
initially start with the CSP rules of the parent. After
that, all new CSP headers should only be set on the
policy container of the iframe.

Part of #36437

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52194).

jdm commented on components/script/script_thread.rs:
```diff
@@ -3708,6 +3712,7 @@ impl ScriptThread {
             dummy_request_id,
             Ok(ResourceFetchTiming::new(ResourceTimingType::None)),
         );
+        context.append_parent_to_csp_list(policy_container.as_ref());
```
It occurs to me that we should move this right after the process_response call, otherwise the srcdoc contents could initiate some blockable operations before we set up the CSP rules.

TimvdLippe replied:
Fixed 👍

TimvdLippe commented:
I was able to workaround the corruption with a fresh `git clone`. Try run is running: https://github.com/TimvdLippe/servo/actions/runs/14781830192

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52194).

TimvdLippe commented:
Seems like my git diff was still corrupt. Will fix that issue tomorrow.

TimvdLippe commented on components/script/dom/servoparser/mod.rs:
```diff
@@ -813,6 +814,27 @@ impl ParserContext {
+                appended_csp_list.to_owned()
+            },
+        };
+        parser.document.set_csp_list(Some(new_csp_list));
```
I also wonder how we should handle the csp_list on both document and global. Here, we set it on the document. But in many of the checks, we use the window global to retrieve the policy container. Are they shared in some way? Shouldn't we have only 1 CSP list per realm, thus the window and document one should point to the same thing?

jdm replied:
The global getter just fetches the container from the appropriate place according to the kind of global it is, so we get the same result if we obtain it from the global or the global's document.

TimvdLippe commented:
Compilation error should be fixed. Another try run: https://github.com/TimvdLippe/servo/actions/runs/14790420352

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52194).

servo-wpt-sync commented:
📝 Transplanted new upstreamable changes to existing upstream WPT pull request (web-platform-tests/wpt#52194).

servo-wpt-sync commented:
🤖 This change no longer contains upstreamable changes to WPT; closed existing upstream pull request (web-platform-tests/wpt#52194).
